chore(templates): Make all IAM policies have a consistent Version declaration and avoid the outdated 2008-10-17 policy language

rclinard-amzn opened the pull request:
This ensures that our CloudFormation templates use the 2012-10-17 IAM policy version.

Since 2008-10-17 is the default policy version, this also adds the Version key to the policies that were missing it.

Based on https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_version.html, the 2008-10-17 policy should not be used for new policies.

This should not change any behavior, but is likely good practice. Since customers might use Copilot's generated CloudFormation templates as a reference, we want to make sure our generated templates follow best practices recommended by other AWS services.

<!-- Provide summary of changes -->

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, 77" -->

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

rclinard-amzn committed:
chore(templates): do not use outdated 2008-10-17 IAM policy version

This ensures that our CloudFormation templates use the 2012-10-17 IAM policy version.

Since 2008-10-17 is the default policy version, this also adds the Version key to the policies that were missing it.

Based on https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_version.html, the 2008-10-17 policy should not be used for new policies.

This should not change any behavior, but is likely good practice.

paragbhingre commented on internal/pkg/deploy/cloudformation/stack/testdata/workloads/windows-svc-test.stack.yml:
```diff
@@ -108,6 +108,7 @@ Resources: # If a bucket URL is specified, that means the template exists.
     Type: AWS::IAM::Role
     Properties:
       AssumeRolePolicyDocument:
+        Version: '2012-10-17'
```
Nice, thanks for adding it to the test data as well :)

paragbhingre commented on internal/pkg/deploy/cloudformation/stack/testdata/workloads/svc-staging.stack.yml:
```diff
@@ -117,6 +117,7 @@ Resources: # If a bucket URL is specified, that means the template exists.
     Type: AWS::IAM::Role
     Properties:
       AssumeRolePolicyDocument:
+        Version: '2012-10-17'
```
Now that I think I curious how these tests were passing without `Version` before? 
Isn't our actual code generates cfn template that has `Version`? interesting... 

No wonder so many integ tests are failing

rclinard-amzn replied:
It looks like having / not having Version was synced between the templates and tests, so they passed because the expected output had the Version data based on where it occurred in the actual output. The integration test failures happened since I only looked at 100 policies out of 290 policies the first time around, which I fixed in the most recent commit.

rclinard-amzn committed:
chore(templates): do not use outdated 2008-10-17 IAM policy version, take 2

I missed a bunch of policies since IntelliJ only shows 100 results by default. There are closer to 300 policies.

The following regex was used to check for missing versions: :\s*Statement

The following regex was used to to count the number of correctly versioned policies: Version: (["']?)2012-10-17(["']?)\s*Statement

rclinard-amzn committed:
chore(templates): Use consistent form of IAM policy language version

A bare `Version: 2012-10-17` directive is parsed as a Date, but the Version property is expected to be a String. CloudFormation handles this properly, but it's still likely bad form.

This commit makes the policy version field consistently a string - single quotes are used because we don't need escape code behavior.

rclinard-amzn commented:
This PR has been updated so that now every IAM policy in the CF template YML files has a consistent `Version: '2012-10-17'` declaration. This avoids the version being parsed as a date, even though CF seems to handle this properly.

rclinard-amzn committed:
chore(templates): Use consistent form of IAM policy language version, take 2

The JetBrains find window strikes again.

iamhopaul123 commented on e2e/multi-pipeline/s3template.yml:
```diff
@@ -29,7 +29,7 @@ Resources:
     Properties:
       PolicyDocument:
-        Version: 2012-10-17
+        Version: '2012-10-17'
```
Why do we add `''` 🤔

rclinard-amzn replied:
Without the quotes, the version is parsed as a Date instead of a String. This didn't seem quite right to me, since the policy language version is defined to be a string, even though CloudFormation appears to handle this properly.

This just makes it so that all of the policy language declarations are consistently defined with single quotes to avoid any of that weirdness.

mergify[bot] committed:
Merge branch 'mainline' into fix/cf-template/iam-version
